webapp/src/ts/services/create-user-for-contacts.service.ts

```diff
@@ -0,0 +1,119 @@
+      throw new Error('The original contact could not be found when replacing the user.');
+    }
+    if (!newContact) {
+      throw new Error('The new contact could not be found when replacing the user.');
+    }
+    if (!originalContact.parent?._id || originalContact.parent._id !== newContact.parent._id) {
+      throw new Error('The new contact must have the same parent as the original contact when replacing a user.');
+    }
+    const originalUsername = this.getCurrentUsername();
+    if (!originalUsername) {
+      throw new Error('The current username could not be found when replacing the user.');
+    }
+
+    const status = this.sessionService.isOnlineOnly() ? UserCreationStatus.READY : UserCreationStatus.PENDING;
+    if (!originalContact.user_for_contact) {
+      originalContact.user_for_contact = {};
+    }
+    if (!originalContact.user_for_contact.replace) {
+      originalContact.user_for_contact.replace = {};
+    }
+    originalContact.user_for_contact.replace[originalUsername] = {
+      status,
+      replacement_contact_id: newContact._id,
+    };
+  }
+
+  isBeingReplaced(contact) {
+    const replacement = contact.user_for_contact?.replace?.[this.getCurrentUsername()];
+    return !!replacement && [UserCreationStatus.PENDING, UserCreationStatus.READY].includes(replacement.status);
+  }
+
+  getReplacedBy(contact): string | undefined {
+    return contact.user_for_contact?.replace?.[this.getCurrentUsername()]?.replacement_contact_id;
+  }
+
+  private getCurrentUsername() {
+    if (!this.currentUsername) {
+      this.currentUsername = this.sessionService.userCtx()?.name;
+    }
+
+    return this.currentUsername;
+  }
+
+  private getReplacedStatus(contact): UserCreationStatus {
+    return UserCreationStatus[contact.user_for_contact.replace[this.getCurrentUsername()].status];
+  }
+
+  private setReplacedStatus(contact, status: UserCreationStatus) {
+    contact.user_for_contact.replace[this.getCurrentUsername()].status = status;
+  }
+
+  private async syncStatusChanged({ to, from }: any) {
+    if (to !== SyncStatus.Success || from !== SyncStatus.Success) {
+      return;
+    }
+
+    const contact = await this.userContactService.get({ hydrateLineage: false });
+    if (!contact || !this.isBeingReplaced(contact)) {
+      return;
+    }
+
+    const status = this.getReplacedStatus(contact);
+    if (status !== UserCreationStatus.PENDING) {
+      return;
+    }
+    // After a user is replaced, the original contact will have status of PENDING.
+    // Set to READY to trigger Sentinel to create a new user (now that all docs are synced).
+    this.setReplacedStatus(contact, UserCreationStatus.READY);
+    await this.dbService
+      .get()
+      .put(contact);
+    // Make sure there is not an ongoing sync before pushing changes to contact
+    if (this.dbSyncService.isSyncInProgress()) {
+      await this.dbSyncService.sync();
+    }
+    await this.dbSyncService.sync(true);
+    await this.sessionService.logout();
```

Comment:
I agree that this is not the best user experience.  However if I am remembering correctly, I believe we decided that we were okay with this behavior in the MVP of this feature (correct me if I am wrong @mrjones-plip @m5r).  

Having the webapp react to the Sentinel changes would solve some of the rough spots around the user being logged in on multiple devices too (since in theory all the devices would know to logout).  But, I think this behavior would require considerable additional code/tests to listen for changes in the user's contact and react to them.  Not sure if we want to do that work now....

Reply:
Also important to note that in an earlier iteration, everything happened within the sentinel transition and it required the user to manually sync twice. One sync for the app to push the changes and another sync for the app to pick up the reparented bits and log the user out.  
It led to a poor experience and many people inside and outside Medic raised their concerns about it.

The current solution isn't perfect either but it doesn't need the user to sync more than once and there is no chance for reports to get lost by being parented to an old contact or by not being synced to the server before the log out.